libbeat/esleg/eslegclient/connection.go

```diff
@@ -152,14 +153,15 @@ func NewConnection(s ConnectionSettings) (*Connection, error) {
 		if err != nil {
 			return nil, err
 		}
-		logp.Info("kerberos client created")
+		logger.Info("kerberos client created")
 	}
 
 	conn := Connection{
 		ConnectionSettings: s,
 		HTTP:               esClient,
 		Encoder:            encoder,
 		log:                logger,
+		responseBuffer:     bytes.NewBuffer(nil),
```

Comment:
I think this field needs to be protected by a mutex.

Let's say two goroutines, G1 and G2 are running concurrently, and both end up calling `conn.execHTTPRequest` at some point in their execution.  Now imagine this sequence of events:

1. G1 copies `resp.Body` into `conn.responseBuffer` (line 458).
2. G2 calls `conn.responseBuffer.Reset()` (line 457).  
3. G1 returns `conn.responseBuffer.Bytes()` (line 468).

The result will be that G1 will return an empty slice of bytes.

Reply:
You bring a valid concern here. Before going in and adding a mutex which is potentially defeating what we want to achieve and actually slowing down the process. 
I might be wrong, but this is the assumption I had, each go routine would have its own connection. One connection would not be shared between 2 go routines, if 2 goroutines needed a connection to the ES they would create a new connection client thus a new buffer. 
If we absolutely need to add a mutex, I would rather change the implementation because if there is a case that 2 go routines shared the same connection then adding a mutex would be similar to not having go routines in the first place.

Reply:
Yes, you're right that _as it stands today_ the only way two goroutines could concurrently call the `execHTTPRequest` method is if they were sharing the same `Connection` object. And from what I can tell, that is not happening anywhere in the code.

Besides checking this by eyeballing the code paths that lead to the `execHTTPRequest` method, I also ran `go test -race ./libbeat/... -count 1 | grep -i 'data race' | wc -l` from the root folder of the Beats repo, first in the `main` branch and then in this PR's branch.  In both cases I got the same results, which is an quick sanity check that the changes in this PR are most likely not introducing additional data races:

#### Running in `main`

```
$ go test -race ./libbeat/... -count 1 | grep -i 'data race' | wc -l
      22
```

#### Running in this PR's branch

```
$ go test -race ./libbeat/... -count 1 | grep -i 'data race' | wc -l
      22
```

---

However, _in the future_, code may be written such that two concurrently-executing goroutines are sharing the same `Connection` object.  

The safest thing to do here would be to add a mutex, but as you noted, this will likely cut into the performance gains we are seeing from the change in this PR.  

The next best thing to do, IMO, would be to document clearly in a comment above the `execHTTPRequest` method AND in comments above each of the constructor functions that return a new `Connection` — `NewConnection`, `NewClients`, and `NewConnectedClient` — AND in a comment above the `Connection` struct itself that this method/struct is not threadsafe.  You could add a link in your comments to this discussion here on GitHub in case someone wants to know all the details.

Reply:
❤️ for the help here, I agree that race conditions are something to be looked-at at in different work streams.

> However, in the future, code may be written such that two concurrently-executing goroutines are sharing the same Connection object.

I think that once we move away from the esleg and to a different output client, this part of the code should not be used anymore.

I added some comment about the thread-safety, please take a look at f181725

Reply:
Comments LGTM.  Thanks @alexsapran!